src/System.Net.Http/tests/StressTests/HttpStress/ClientOperations.cs

```diff
@@ -23,60 +24,73 @@ public sealed class RequestContext
-            MaxRequestUriSize = maxRequestUriSize;
-            ContentSource = contentSource;
+            MaxRequestParameters = config.MaxParameters;
+            MaxRequestUriSize = config.MaxRequestUriSize;
+            MaxContentLength = config.MaxContentLength;
         }
 
         public int TaskNum { get; }
+        public Version HttpVersion { get; }
         public bool IsCancellationRequested { get; set; }
-        public string ContentSource { get; }
         public int MaxRequestParameters { get; }
         public int MaxRequestUriSize { get; }
-        public int MaxContentLength => ContentSource.Length;
+        public int MaxContentLength { get; }
         public Uri BaseAddress => _client.BaseAddress;
 
         // HttpClient.SendAsync() wrapper that wires randomized cancellation
         public async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption httpCompletion = HttpCompletionOption.ResponseContentRead, CancellationToken? token = null)
         {
+            request.Version = HttpVersion;
+
             if (token != null)
             {
                 // user-supplied cancellation token overrides random cancellation
-                return await _client.SendAsync(request, httpCompletion, token.Value);
+                using var cts = CancellationTokenSource.CreateLinkedTokenSource(_globalToken, token.Value);
+                return await _client.SendAsync(request, httpCompletion, cts.Token);
             }
             else if (GetRandomBoolean(_cancellationProbability))
             {
                 // trigger a random cancellation
-                using (var cts = new CancellationTokenSource())
+                using var cts = CancellationTokenSource.CreateLinkedTokenSource(_globalToken);
+
+                Task<HttpResponseMessage> task = _client.SendAsync(request, httpCompletion, cts.Token);
+
+                // either spinwait or delay before triggering cancellation
+                if (GetRandomBoolean(probability: 0.66))
                 {
-                    int delayMs = _random.Next(0, 2);
-                    Task<HttpResponseMessage> task = _client.SendAsync(request, httpCompletion, cts.Token);
-                    if (delayMs > 0)
-                        await Task.Delay(delayMs);
-
-                    cts.Cancel();
-                    IsCancellationRequested = true;
-                    return await task;
+                    // bound spinning to 100 us
+                    double spinTimeMs = 0.1 * _random.NextDouble();
+                    Stopwatch sw = Stopwatch.StartNew();
+                    do { Thread.SpinWait(10); } while (!task.IsCompleted && sw.Elapsed.TotalMilliseconds < spinTimeMs);
                 }
+                else
+                {
+                    // 60ms is the 99th percentile when
+                    // running the stress suite locally under default load
+                    await Task.WhenAny(task, Task.Delay(_random.Next(0, 60)));
```

Comment:
This has the potential to create quite a lot of timers that go undisposed.  Since you're going to always cancel the `cts` anyway, can you pass `cts.Token` as an argument to Task.Delay?  That way, if we complete the WhenAny because `task` completed, we'll cancel the timer (and if it's the timer that causes the WhenAny to complete, canceling it will be a nop).

Reply:
Done